File: client-py/queen/builders/operation_builder.py
"""
Operation builder for create/delete operations with callbacks
"""

from typing import Any, Callable, Dict, Optional

from ..utils import logger


class OperationBuilder:
    """Operation builder for create/delete operations with callbacks"""

    def __init__(self, http_client: Any, method: str, path: str, body: Optional[Dict[str, Any]]):
        """
        Initialize operation builder

        Args:
            http_client: HttpClient instance
            method: HTTP method
            path: Request path
            body: Request body
        """
        self._http_client = http_client
        self._method = method
        self._path = path
        self._body = body
        self._on_success_callback: Optional[Callable[[Any], Any]] = None
        self._on_error_callback: Optional[Callable[[Exception], Any]] = None
        self._executed = False

    def on_success(self, callback: Callable[[Any], Any]) -> "OperationBuilder":
        """
        Set success callback

        Args:
            callback: Success callback

        Returns:
            Self for chaining
        """
        self._on_success_callback = callback
        return self

    def on_error(self, callback: Callable[[Exception], Any]) -> "OperationBuilder":
        """
        Set error callback

        Args:
            callback: Error callback

        Returns:
            Self for chaining
        """
        self._on_error_callback = callback
        return self

    def __await__(self):
        """Make awaitable"""
        return self._execute().__await__()

    async def _execute(self) -> Any:
        """Execute operation"""
        if self._executed:
            return None
        self._executed = True

        logger.log("OperationBuilder.execute", {"method": self._method, "path": self._path})

        try:
            if self._method == "GET":
                result = await self._http_client.get(self._path)
            elif self._method == "POST":
                result = await self._http_client.post(self._path, self._body)
            elif self._method == "PUT":
                result = await self._http_client.put(self._path, self._body)
            elif self._method == "DELETE":
                result = await self._http_client.delete(self._path)
            else:
                raise ValueError(f"Unsupported method: {self._method}")

            if result and isinstance(result, dict) and result.get("error"):
                error = Exception(result["error"])
                logger.error(
                    "OperationBuilder.execute",
                    {"method": self._method, "path": self._path, "error": result["error"]},
                )
                if self._on_error_callback:
                    await self._on_error_callback(error)
                    return {"success": False, "error": result["error"]}
                raise error

            logger.log("OperationBuilder.execute", {"method": self._method, "path": self._path, "status": "success"})
            if self._on_success_callback:
                await self._on_success_callback(result)

            return result

        except Exception as error:
            logger.error("OperationBuilder.execute", {"method": self._method, "path": self._path, "error": str(error)})
            if self._on_error_callback:
                await self._on_error_callback(error)
                return {"success": False, "error": str(error)}
            raise

File: client-laravel/src/Builders/OperationBuilder.php
<?php

namespace Queen\Builders;

use Queen\Http\HttpClient;

class OperationBuilder
{
    private HttpClient $httpClient;
    private string $method;
    private string $path;
    private ?array $body;
    private ?\Closure $onSuccessCallback = null;
    private ?\Closure $onErrorCallback = null;

    public function __construct(HttpClient $httpClient, string $method, string $path, ?array $body)
    {
        $this->httpClient = $httpClient;
        $this->method = $method;
        $this->path = $path;
        $this->body = $body;
    }

    public function onSuccess(\Closure $callback): static
    {
        $this->onSuccessCallback = $callback;
        return $this;
    }

    public function onError(\Closure $callback): static
    {
        $this->onErrorCallback = $callback;
        return $this;
    }

    public function execute(): mixed
    {
        try {
            $result = match ($this->method) {
                'GET' => $this->httpClient->get($this->path),
                'POST' => $this->httpClient->post($this->path, $this->body),
                'PUT' => $this->httpClient->put($this->path, $this->body),
                'DELETE' => $this->httpClient->delete($this->path),
            };

            if (is_array($result) && isset($result['error'])) {
                $error = new \RuntimeException($result['error']);
                if ($this->onErrorCallback !== null) {
                    ($this->onErrorCallback)($error);
                    return ['success' => false, 'error' => $result['error']];
                }
                throw $error;
            }

            if ($this->onSuccessCallback !== null) {
                ($this->onSuccessCallback)($result);
            }

            return $result;
        } catch (\Throwable $error) {
            if ($this->onErrorCallback !== null) {
                ($this->onErrorCallback)($error);
                return ['success' => false, 'error' => $error->getMessage()];
            }
            throw $error;
        }
    }
}
